src/items/functions.md

```diff
@@ -203,12 +203,30 @@ extern "C" fn new_i32() -> i32 { 0 }
 let fptr: extern "C" fn() -> i32 = new_i32;
 ```
 
-Functions with an ABI that differs from `"Rust"` do not support unwinding in the
-exact same way that Rust does. Therefore, unwinding past the end of functions
-with such ABIs causes the process to abort.
+### Unwinding
+
+Each non-`Rust` ABI string has a corresponding ABI with the suffix `-unwind`.
+
+The choice of ABI, together with the [panic mode][panic-modes], determines the
+behavior when unwinding out of a function.
+
+In the table below, "Unforced foreign unwind" refers to something like a C++
+exception; the table indicates the behavior when entering a Rust stackframe via
+an function or function pointer declared with the specified ABI string.
+Additionally, `C` in the ABI strings may be substituted with `stdcall` or any
+other ABI supported by the language implementation.
+
+| panic runtime  | ABI          | `panic`-unwind                        | Unforced foreign unwind |
+| -------------- | ------------ | ------------------------------------- | ----------------------- |
+| `panic=unwind` | `"C-unwind"` | unwind                                | unwind                  |
+| `panic=unwind` | `"C"`        | abort                                 | UB                      |
+| `panic=abort`  | `"C-unwind"` | `panic!` aborts                       | abort                   |
+| `panic=abort`  | `"C"`        | `panic!` aborts (no unwinding occurs) | UB                      |
```

Comment:
```suggestion
| panic runtime  | ABI          | `panic`-unwind                        | Unforced foreign unwind |
| -------------- | ------------ | ------------------------------------- | ----------------------- |
| `panic=unwind` | `"C-unwind"` | unwind                                | unwind                  |
| `panic=unwind` | `"C"`        | abort if unwinding reaches the function                                | UB if unwinding reaches the function                      |
| `panic=abort`  | `"C-unwind"` | aborts immediately (no unwinding occurs)                      | abort if unwinding reaches the function                   |
| `panic=abort`  | `"C"`        | aborts immediately (no unwinding occurs) | UB if unwinding reaches the function                      |
```

I found this a bit confusing. I believe there are subtle differences in terms of where the aborts occur and so forth. I have tried to clarify above, but I think it may be worth further clarifying.

It may also be worth adding some (perhaps non-normative) discussion of implementation:

* When compiling a function F with `panic=unwind` and `extern "C"`, the compiler inserts unwinding guards for Rust panics that trigger an abort when unwinding reaches F. 

I am also be misunderstanding what's going on. I was a bit surprised to see "UB" for unforced-foreign-unwind with C=unwind. I guess that this table is combining two scenarios:

* what happens when you call a C++ function declared as extern "C", and it unwinds (UB, we haven't compiled any guards)
* what happens when an `extern "C"` Rust function invokes some C++ function that throws (probably, in practice, an abort, but perhaps we have simplified to call it UB?)

Is that right?

Reply:
It's only UB for a foreign function declared as `extern "C"` to unwind.

Reply:
@nbdd0121 what happens when an `extern "C"` Rust function unwinds? I believe we insert an abort guard, but this table doesn't clarify that, right? Or maybe I don't understand what it's trying to convey. I'm imagining a scenario like

```rust
extern "C-unwind" fn throws();

extern "C" fn rust_fn() {
    throws(); // unwinds
}
```

In this case, I presume you get an abort -- and I think we guarantee that? But the way I read this table, it would be listed as UB.

Reply:
Hm....I don't know if the `panic` abort guard would currently catch and abort in that case, or if it relies on the personality function to only abort on true Rust `panic`s. I agree that the behavior in the table as-written is UB.

Reply:
> @nbdd0121 what happens when an `extern "C"` Rust function unwinds? I believe we insert an abort guard, but this table doesn't clarify that, right? Or maybe I don't understand what it's trying to convey. I'm imagining a scenario like
> 
> ```rust
> extern "C-unwind" fn throws();
> 
> extern "C" fn rust_fn() {
>     throws(); // unwinds
> }
> ```
> 
> In this case, I presume you get an abort -- and I think we guarantee that? But the way I read this table, it would be listed as UB.

Unwinding out from `extern "C"` functions (defined in either Rust or foreign language) is UB.
In the case you listed, we insert guard to prevent unwinding from actually leaving a Rust `extern "C"` functions, therefore the function does not unwind, so UB is prevented; in this case we never unwinds out from a `extern "C"` Rust functions.

If you define a `extern "C-unwind"` Rust function and transmute it to `extern "C"` and then call it, it's not UB if unwinding does not happen, and it's UB if unwinding happens.

Reply:
@nikomatsakis  With the change to the verbiage above, explaining that the table entries are specifically describing behavior at function boundaries, do you still want to make a change here?

Reply:
Please check whether the notes I suggested to add under the table are correct.

Reply:
@nikomatsakis Can I resolve this comment thread now?